app/src/main/java/de/vectordata/skynet/net/PacketHandler.java

```diff
@@ -365,8 +338,58 @@ public void handlePacket(P33DeviceListResponse packet) {
 
     }
 
+    ////////////////////// Utility methods //////////////////////
+    private void regenerateLoopbackKeys(Channel loopbackChannel, Channel accountDataChannel) {
+        Log.d(TAG, "Incomplete or missing loopback channel keys. Regenerating...");
+        Storage.getDatabase().channelKeyDao().dropKeys(loopbackChannel.getChannelId());
+
+        EC.KeyMaterial signature = EC.generateKeypair();
+        EC.KeyMaterial derivation = EC.generateKeypair();
+        if (signature == null || derivation == null)
+            return;
+
+        P17PrivateKeys privateKeys = new P17PrivateKeys(signature.getPrivateKey(), derivation.getPrivateKey());
+        P18PublicKeys publicKeys = new P18PublicKeys(signature.getPublicKey(), derivation.getPublicKey());
+
+        // First transmit the private keys
+        SkynetContext.getCurrent().getMessageInterface().send(loopbackChannel.getChannelId(), privateKeys)
+                .waitForSuccess(r -> {
+                    // If that was successful, store them in the database
+                    Storage.getDatabase().channelKeyDao().insert(ChannelKey.fromPacket(privateKeys));
```

Comment:
This insert operation leads to foreign key constraint violation. I would assume that the channel message entity is missing.

```
android.database.sqlite.SQLiteConstraintException: FOREIGN KEY constraint failed (code 787 SQLITE_CONSTRAINT_FOREIGNKEY)
        at android.database.sqlite.SQLiteConnection.nativeExecuteForLastInsertedRowId(Native Method)
        at android.database.sqlite.SQLiteConnection.executeForLastInsertedRowId(SQLiteConnection.java:796)
        at android.database.sqlite.SQLiteSession.executeForLastInsertedRowId(SQLiteSession.java:788)
        at android.database.sqlite.SQLiteStatement.executeInsert(SQLiteStatement.java:86)
        at androidx.sqlite.db.framework.FrameworkSQLiteStatement.executeInsert(FrameworkSQLiteStatement.java:51)
        at androidx.room.EntityInsertionAdapter.insert(EntityInsertionAdapter.java:80)
        at de.vectordata.skynet.data.sql.dao.ChannelKeyDao_Impl.insert(ChannelKeyDao_Impl.java:127)
        at de.vectordata.skynet.net.PacketHandler.lambda$regenerateLoopbackKeys$1(PacketHandler.java:358)
        at de.vectordata.skynet.net.-$$Lambda$PacketHandler$bBYhDRvXKwqnJlbPzc31tSdyohI.handle(Unknown Source:8)
        at de.vectordata.skynet.net.response.PacketTask.lambda$waitForSuccess$1(PacketTask.java:50)
        at de.vectordata.skynet.net.response.-$$Lambda$PacketTask$Sl6VMANmQM1q2YYbZeKo49bevp8.handle(Unknown Source:4)
        at de.vectordata.skynet.net.response.PacketTask.lambda$waitForResponse$2$PacketTask(PacketTask.java:68)
        at de.vectordata.skynet.net.response.-$$Lambda$PacketTask$m69dEHnf4uwKOAfbzFBGtf9K5lc.handle(Unknown Source:8)
        at de.vectordata.skynet.net.response.BaseAwaiterItem.handle(BaseAwaiterItem.java:18)
        at de.vectordata.skynet.net.response.PacketTask.onPacketEvent(PacketTask.java:79)
        at java.lang.reflect.Method.invoke(Native Method)
        at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:507)
        at org.greenrobot.eventbus.EventBus.postToSubscription(EventBus.java:430)
        at org.greenrobot.eventbus.EventBus.postSingleEventForEventType(EventBus.java:411)
        at org.greenrobot.eventbus.EventBus.postSingleEvent(EventBus.java:384)
        at org.greenrobot.eventbus.EventBus.post(EventBus.java:265)
        at de.vectordata.skynet.net.PacketHandler.handlePacket(PacketHandler.java:121)
        at de.vectordata.skynet.net.NetworkManager.onPacketReceived(NetworkManager.java:136)
        at de.vectordata.skynet.net.client.SslClient.receiveLoop(SslClient.java:74)
        at de.vectordata.skynet.net.client.SslClient.lambda$connect$0$SslClient(SslClient.java:41)
        at de.vectordata.skynet.net.client.-$$Lambda$SslClient$seCGbRtAjmtpta9VGd0U12rFxRs.run(Unknown Source:6)
        at java.lang.Thread.run(Thread.java:764)
```

Reply:
did you reset the database and use a new server/accounmt, the old client broke all loopback channels?

Reply:
Please see latest commit for a fix